README.md
PicoModal [![Build Status](https://secure.travis-ci.org/Nycto/PicoModal.png?branch=master)](http://travis-ci.org/Nycto/PicoModal) [![Bower version](https://badge.fury.io/bo/picoModal.png)](http://badge.fury.io/bo/picoModal) [![npm version](https://badge.fury.io/js/picomodal.svg)](https://www.npmjs.com/package/picomodal)
=========

A small, self-contained JavaScript modal library

* Small: At around 2kb minified & gzipped, it's small and easily embeddable
* No Dependencies: PicoModal does not depend on any other JS libraries,
  so you can use it in places where you don't have access to one
* Self-contained: No extra CSS or images required; just the JS
* Simple: The interface is straight forward and easy to use
* Customizable: By changing a few settings you can customize or completely
  replace the default styles and behaviour
* Accessible: Handles focus management, keyboard events and Aria tags

Download
--------

The latest version of PicoModal is available here:
[Download](https://github.com/Nycto/PicoModal/releases)

Browser Support
---------------

![Browser Support Matrix](https://saucelabs.com/browser-matrix/picomodal.svg?auth=97d19e9c77a6980f249c9e6cf7ab78ec)

Basic Example
-------------

If all you want to do is display a modal, it's as easy as this:
([Run this code](http://jsfiddle.net/nnrXD/))

```javascript
  picoModal("Ah, the pitter patter of tiny feet in huge combat boots.").show();
```

If you plan on showing the same modal multiple times, make sure you keep a
reference to the instance, like this:
([Run this code](http://jsfiddle.net/2s95h/))

```javascript
  var modal = picoModal("Ah, the pitter patter of tiny feet in huge combat boots.");
  document.getElementById("modal").addEventListener("click", function(){
      modal.show();
  });
```

For more control over the behaviour of the modal, you can pass in a settings
object: ([Run this code](http://jsfiddle.net/px7LU/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      overlayStyles: {
          backgroundColor: "#169",
          opacity: 0.75
      }
  }).show();
```

A full list of settings is documented below.

Manually Closing a Modal
------------------------

If you want to programatically close the modal you can do it like this:
([Run this code](http://jsfiddle.net/KkV27/))

```javascript
  var modal = picoModal(
      "<p>Ah, the pitter patter of tiny feet in huge combat boots.<p>"
      + "<p><a href='#' class='dismiss'>Dismiss</a></p>"
  ).show();

  document.body.addEventListener('click', function(event) {
    if( /\bdismiss\b/.test(event.target.className) ) {
      modal.close();
    }
  });
```

Or you can use a more targetted implementation with the `afterCreate` event:
([Run this code](http://jsfiddle.net/GZb2z/))

```javascript
  picoModal(
      "<p>Ah, the pitter patter of tiny feet in huge combat boots.<p>"
      + "<p><a href='#' class='dismiss'>Dismiss</a></p>"
  ).afterCreate(function(modal){
      modal.modalElem().getElementsByClassName("dismiss")[0]
          .addEventListener('click', modal.close);
  }).show();
```

You can also attach an event to fire when the modal is closed:
([Run this code](http://jsfiddle.net/Lkx5r/))

```javascript
  picoModal("Ah, the pitter patter of tiny feet in huge combat boots.")
      .afterClose(function () { alert("Closed"); })
      .show();
```

Customizing Behavior
--------------------

To disable the close button, and instead just rely on someone clicking
outside of the modal, you can do this:
([Run this code](http://jsfiddle.net/WFqz4/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      closeButton: false
  }).show();
```

Or, to disable closing when someone clicks outside of the modal, you can
do this: ([Run this code](http://jsfiddle.net/7d53v/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      overlayClose: false
  }).show();
```

To use custom HTML for the close button, do this:
([Run this code](http://jsfiddle.net/UvcDL/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      closeHtml: "<span>Close</span>",
      closeStyles: {
          position: "absolute", top: "-10px", right: "-10px",
          background: "#eee", padding: "5px 10px", cursor: "pointer",
          borderRadius: "5px", border: "1px solid #ccc"
      }
  }).show();
```

Events
------

There are a few events you can hook into for watching and sometimes monitoring
the behavior of a modal. The events are:

* `afterCreate`: Triggered when the DOM Nodes for a modal are created
* `beforeShow`: Triggered before a modal is shown. Allows for cancellation
* `afterShow`: Triggered after a modal is shown
* `beforeClose`: Triggered before a modal is closed. Allows for cancellation
* `afterClose`: triggered after a modal is closed

These exist as methods on the PicoModal instance. You can use them like this:
([Run this code](http://jsfiddle.net/875Tx/))

```javascript
  picoModal("Ah, the pitter patter of tiny feet in huge combat boots.")
      .afterClose(function (modal) {
          alert("Modal Closed: " + modal.modalElem().innerText);
      })
      .show();
```

The first argument passed to the callback is the PicoModal instance for the
specific modal.

For two of the events, beforeShow and beforeClose, there is a second argument
passed that lets you cancel the behavior in question. For example:
([Run this code](http://jsfiddle.net/m83nW/))

```javascript
  picoModal("Ah, the pitter patter of tiny feet in huge combat boots.")
      .beforeShow(function (modal, event) {
          if ( !confirm("Are you sure you want to open this modal?") ) {
              event.preventDefault();
          }
      })
      .show();
```

Single Shot Modal
-----------------

You can use the `afterClose` event and the `destroy` method to create a modal
that will clean up after itself when it is closed, like this:
([Run this code](http://jsfiddle.net/wxZ56/))

```javascript
  picoModal("Ah, the pitter patter of tiny feet in huge combat boots.")
      .afterClose(function (modal) { modal.destroy(); })
      .show();
```

Animation
---------

PicoModal doesn't have any built in animations, but you can use the event
system to add some of your own. For example, the following snippet adds a fade
in and out using jQuery: ([Run this code](http://jsfiddle.net/Sr697/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      overlayStyles: function ( styles ) { styles.opacity = 0; },
      modalStyles: function ( styles ) { styles.opacity = 0; }
  })
  .afterShow(function(modal){
      $(modal.overlayElem()).animate({opacity: .5});
      $(modal.modalElem()).animate({opacity: 1});
  })
  .beforeClose(function(modal, event) {
      event.preventDefault();
      $(modal.overlayElem()).add(modal.modalElem())
          .animate(
              { opacity: 0 },
              { complete: modal.forceClose }
          );
  })
  .show();
```

Settings
--------

The following settings are available when creating a modal:

* __content__: The data to display to the user
* __width__: The forced width of the modal
* __closeButton__: Boolean whether to display the close button
* __closeHtml__: Custom HTML content for the close button
* __closeStyles__: A hash of CSS properties to apply to the close button
* __closeClass__: A class to attach to the close button
* __overlayClose__: Boolean whether a click on the shadow should close the modal
* __overlayStyles__: A hash of additional CSS properties to apply to the
  overlay behind the modal
* __overlayClass__: A class to attach to the overlay element
* __modalStyles__: A hash of additional CSS properties to apply to the
  modal element
* __modalClass__: A class to attach to the main modal element
* __modalId__: The ID to assign to the modal element. A default ID is generated
  used if none is provided.
* __parent__: By default, the modal dialog elements are attached to
  `document.body`. This options allows you to select an alternative parent
  element by specifying a node or a selector
* __escCloses__: When `false`, disables pressing the escape key to close this
  modal. This defaults to `true`.
* __focus__: Whether to automatically set focus on the first focusable element
  within this modal when it opens. This defaults to `true`.
* __ariaDescribedBy__: The id of the element that contains the main content
  of this modal. This sets the `[aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute)`
  attribute. This defaults to the ID of the modal if none is provided.
* __ariaLabelledBy__: The id of the element that contains the general label
  for this modal. This sets the `[aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute)`
  attributed. It is left blank if none is provided.
* __bodyOverflow__: Whether to set `overflow: hidden` on the body when the modal
  is displayed. This prevents the main page from scrolling when a modal is open

If a method is passed as an argument for any of the settings, it will be
called. The first argument passed in is the default value for that setting. This
makes it easy to modify the defaults instead of having to totally define your
own, like so: ([Run this code](http://jsfiddle.net/rptf6/))

```javascript
  picoModal({
      content: "Ah, the pitter patter of tiny feet in huge combat boots.",
      overlayStyles: function (styles) {
          styles.opacity = 0.1;
          return styles;
      }
  }).show();
```

Modal Instance API
------------------

The following methods are available on the object returned by `picoModal`:

* __modalElem__: Returns the modal DOM Node
* __closeElem__: Returns the close button DOM Node
* __overlayElem__: Returns the overlay DOM Node
* __show__: Displays the modal
* __buildDom__: Builds the DOM for the modal, but without showing it
* __close__: Hides the modal
* __forceClose__: Hides the modal without calling the beforeClose events
* __destroy__: Detaches all DOM Nodes and unhooks this modal
* __isVisible__: Whether this modal is currently being displayed
* __options__: Updates the options for this modal. This will only let you
  change options that are re-evaluted regularly, such as `overlayClose`.
* __afterCreate__: Registers a callback to invoke when the modal is created
* __beforeShow__: Registers a callback to invoke before the modal is shown
* __afterShow__: Registers a callback to invoke when the modal is shown
* __beforeClose__: Registers a callback to invoke before the modal is closed
* __afterClose__: Registers a callback to invoke when the modal is closed

License
-------

PicoModal is released under the MIT License, which is pretty spiffy. You should
have received a copy of the MIT License along with this program. If not, see
http://www.opensource.org/licenses/mit-license.php



CHANGES.md
Version Changelog
=================

3.0.0
-----

* Accessibility
  * Set and return focus on show/hide
  * Tab capturing and looping within the modal
  * Close the modal when `escape` is pressed
  * Aria attributes
* Full unit test support
* Vertically position at 38.1966% of the viewport height
* Max height of 90% of viewport height
* Set `box-sizing: border-box`
* Set `overflow: auto`
* Use a button tag for the close element
* Add the isVisible API method
* Automatically attach an ID to all modals

2.3.0
-----

* Add the `parent` option
* Bug fix for undefined class on the pico-close element

2.2.0
-----

* Browserify support
* Add the `buildDom` API method

2.1.0
-----

* Improved centering of modals via `translate`
* Remove IE8 handlers
* Add Aria dialog role



grunt.js
/*global module:false*/
module.exports = function(grunt) {

    grunt.loadNpmTasks('grunt-string-replace');

    // Project configuration.
    grunt.initConfig({
        pkg: '<json:package.json>',
        lint: {
            files: ['grunt.js', 'lib/**/*.js']
        },
        concat: {
            dist: {
                src: ['<file_strip_banner:lib/picoModal.js>'],
                dest: 'dist/picoModal.concat.js'
            }
        },
        "string-replace": {
            dist: {
                src: [ '<config:concat.dist.dest>' ],
                dest: 'dist/picoModal.replaced.js',
                options: {
                    replacements: [
                        { pattern: /\bwatch\b/g, replacement: "w" },
                        { pattern: /\btrigger\b/g, replacement: "t" },
                        { pattern: /\belem\b/g, replacement: "e" },
                        { pattern: /\bchild\b/g, replacement: "c" },
                        { pattern: /\bstylize\b/g, replacement: "s" },
                        { pattern: /\bclazz\b/g, replacement: "z" },
                        { pattern: /\bhtml\b/g, replacement: "h" },
                        { pattern: /\bgetWidth\b/g, replacement: "d" },
                        { pattern: /\bonClick\b/g, replacement: "o" },
                        { pattern: /\bdestroy\b/g, replacement: "x" }
                    ]
                }
            }
        },
        min: {
            dist: {
                src: ['<config:string-replace.dist.dest>'],
                dest: 'dist/picoModal.min.js'
            }
        },
        watch: {
            files: '<config:lint.files>',
            tasks: 'lint'
        },
        jshint: {},
        uglify: {}
    });

    // Default task.
    grunt.registerTask('default', 'lint concat string-replace min');

};


src/picoModal.js
/**
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

(function (root, factory) {
    "use strict";

    if (typeof define === 'function' && define.amd) {
        define([], factory);
    }
    else if (typeof module === 'object' && module.exports) {
        module.exports = factory();
    }
    else {
        root.picoModal = factory();
    }
}(this, function () {

    /**
     * A self-contained modal library
     */
    "use strict";

    /** Returns whether a value is a dom node */
    function isNode(value) {
        if ( typeof Node === "object" ) {
            return value instanceof Node;
        }
        else {
            return value &&
                typeof value === "object" &&
                typeof value.nodeType === "number";
        }
    }

    /** Returns whether a value is a string */
    function isString(value) {
        return typeof value === "string";
    }

    /**
     * Generates observable objects that can be watched and triggered
     */
    function observable() {
        var callbacks = [];
        return {
            watch: callbacks.push.bind(callbacks),
            trigger: function( context ) {

                var unprevented = true;
                var event = {
                    preventDefault: function preventDefault () {
                        unprevented = false;
                    }
                };

                for (var i = 0; i < callbacks.length; i++) {
                    callbacks[i](context, event);
                }

                return unprevented;
            }
        };
    }


    /** Whether an element is hidden */
    function isHidden ( elem ) {
        // @see http://stackoverflow.com/questions/19669786
        return window.getComputedStyle(elem).display === 'none';
    }


    /**
     * A small interface for creating and managing a dom element
     */
    function Elem( elem ) {
        this.elem = elem;
    }

    /** Creates a new div */
    Elem.make = function ( parent, tag ) {
        if ( typeof parent === "string" ) {
            parent = document.querySelector(parent);
        }
        var elem = document.createElement(tag || 'div');
        (parent || document.body).appendChild(elem);
        return new Elem(elem);
    };

    Elem.prototype = {

        /** Creates a child of this node */
        child: function (tag) {
            return Elem.make(this.elem, tag);
        },

        /** Applies a set of styles to an element */
        stylize: function(styles) {
            styles = styles || {};

            if ( typeof styles.opacity !== "undefined" ) {
                styles.filter =
                    "alpha(opacity=" + (styles.opacity * 100) + ")";
            }

            for (var prop in styles) {
                if (styles.hasOwnProperty(prop)) {
                    this.elem.style[prop] = styles[prop];
                }
            }

            return this;
        },

        /** Adds a class name */
        clazz: function (clazz) {
            this.elem.className += " " + clazz;
            return this;
        },

        /** Sets the HTML */
        html: function (content) {
            if ( isNode(content) ) {
                this.elem.appendChild( content );
            }
            else {
                this.elem.innerHTML = content;
            }
            return this;
        },

        /** Adds a click handler to this element */
        onClick: function(callback) {
            this.elem.addEventListener('click', callback);
            return this;
        },

        /** Removes this element from the DOM */
        destroy: function() {
            this.elem.parentNode.removeChild(this.elem);
        },

        /** Hides this element */
        hide: function() {
            this.elem.style.display = "none";
        },

        /** Shows this element */
        show: function() {
            this.elem.style.display = "block";
        },

        /** Sets an attribute on this element */
        attr: function ( name, value ) {
            if (value !== undefined) {
                this.elem.setAttribute(name, value);
            }
            return this;
        },

        /** Executes a callback on all the ancestors of an element */
        anyAncestor: function ( predicate ) {
            var elem = this.elem;
            while ( elem ) {
                if ( predicate( new Elem(elem) ) ) {
                    return true;
                }
                else {
                    elem = elem.parentNode;
                }
            }
            return false;
        },

        /** Whether this element is visible */
        isVisible: function () {
            return !isHidden(this.elem);
        }
    };


    /** Generates the grey-out effect */
    function buildOverlay( getOption, close ) {
        return Elem.make( getOption("parent") )
            .clazz("pico-overlay")
            .clazz( getOption("overlayClass", "") )
            .stylize({
                display: "none",
                position: "fixed",
                top: "0px",
                left: "0px",
                height: "100%",
                width: "100%",
                zIndex: 10000
            })
            .stylize(getOption('overlayStyles', {
                opacity: 0.5,
                background: "#000"
            }))
            .onClick(function () {
                if ( getOption('overlayClose', true) ) {
                    close();
                }
            });
    }

    // An auto incrementing ID assigned to each modal
    var autoinc = 1;

    /** Builds the content of a modal */
    function buildModal( getOption, close ) {
        var width = getOption('width', 'auto');
        if ( typeof width === "number" ) {
            width = "" + width + "px";
        }

        var id = getOption("modalId", "pico-" + autoinc++);

        var elem = Elem.make( getOption("parent") )
            .clazz("pico-content")
            .clazz( getOption("modalClass", "") )
            .stylize({
                display: 'none',
                position: 'fixed',
                zIndex: 10001,
                left: "50%",
                top: "38.1966%",
                maxHeight: '90%',
                boxSizing: 'border-box',
                width: width,
                '-ms-transform': 'translate(-50%,-38.1966%)',
                '-moz-transform': 'translate(-50%,-38.1966%)',
                '-webkit-transform': 'translate(-50%,-38.1966%)',
                '-o-transform': 'translate(-50%,-38.1966%)',
                transform: 'translate(-50%,-38.1966%)'
            })
            .stylize(getOption('modalStyles', {
                overflow: 'auto',
                backgroundColor: "white",
                padding: "20px",
                borderRadius: "5px"
            }))
            .html( getOption('content') )
            .attr("id", id)
            .attr("role", "dialog")
            .attr("aria-labelledby", getOption("ariaLabelledBy"))
            .attr("aria-describedby", getOption("ariaDescribedBy", id))
            .onClick(function (event) {
                var isCloseClick = new Elem(event.target)
                    .anyAncestor(function (elem) {
                        return /\bpico-close\b/.test(elem.elem.className);
                    });
                if ( isCloseClick ) {
                    close();
                }
            });

        return elem;
    }

    /** Builds the close button */
    function buildClose ( elem, getOption ) {
        if ( getOption('closeButton', true) ) {
            return elem.child('button')
                .html( getOption('closeHtml', "&#xD7;") )
                .clazz("pico-close")
                .clazz( getOption("closeClass", "") )
                .stylize( getOption('closeStyles', {
                    borderRadius: "2px",
                    border: 0,
                    padding: 0,
                    cursor: "pointer",
                    height: "15px",
                    width: "15px",
                    position: "absolute",
                    top: "5px",
                    right: "5px",
                    fontSize: "16px",
                    textAlign: "center",
                    lineHeight: "15px",
                    background: "#CCC"
                }) )
                .attr("aria-label", getOption("close-label", "Close"));
        }
    }

    /** Builds a method that calls a method and returns an element */
    function buildElemAccessor( builder ) {
        return function () {
            return builder().elem;
        };
    }


    // An observable that is triggered whenever the escape key is pressed
    var escapeKey = observable();

    // An observable that is triggered when the user hits the tab key
    var tabKey = observable();

    /** A global event handler to detect the escape key being pressed */
    document.documentElement.addEventListener(
        'keydown',
        function onKeyPress (event) {
            var keycode = event.which || event.keyCode;

            // If this is the escape key
            if ( keycode === 27 ) {
                escapeKey.trigger();
            }

            // If this is the tab key
            else if ( keycode === 9 ) {
                tabKey.trigger(event);
            }
        }
    );


    /** Attaches focus management events */
    function manageFocus ( iface, isEnabled ) {

        /** Whether an element matches a selector */
        function matches ( elem, selector ) {
            var fn = elem.msMatchesSelector ||
                elem.webkitMatchesSelector ||
                elem.matches;
            return fn.call(elem, selector);
        }

        /**
         * Returns whether an element is focusable
         * @see http://stackoverflow.com/questions/18261595
         */
        function canFocus( elem ) {
            if (
                isHidden(elem) ||
                matches(elem, ":disabled") ||
                elem.hasAttribute("contenteditable")
            ) {
                return false;
            }
            else {
                return elem.hasAttribute("tabindex") ||
                    matches(
                        elem,
                        "input,select,textarea,button,a[href],area[href],iframe"
                    );
            }
        }

        /** Returns the first descendant that can be focused */
        function firstFocusable ( elem ) {
            var items = elem.getElementsByTagName("*");
            for (var i = 0; i < items.length; i++) {
                if ( canFocus(items[i]) ) {
                    return items[i];
                }
            }
        }

        /** Returns the last descendant that can be focused */
        function lastFocusable ( elem ) {
            var items = elem.getElementsByTagName("*");
            for (var i = items.length; i--;) {
                if ( canFocus(items[i]) ) {
                    return items[i];
                }
            }
        }

        // The element focused before the modal opens
        var focused;

        // Records the currently focused element so state can be returned
        // after the modal closes
        iface.beforeShow(function getActiveFocus() {
            focused = document.activeElement;
        });

        // Shift focus into the modal
        iface.afterShow(function focusModal() {
            if ( isEnabled() ) {
                var focusable = firstFocusable(iface.modalElem());
                if ( focusable ) {
                    focusable.focus();
                }
            }
        });

        // Restore the previously focused element when the modal closes
        iface.afterClose(function returnFocus() {
            if ( isEnabled() && focused ) {
                focused.focus();
            }
            focused = null;
        });

        // Capture tab key presses and loop them within the modal
        tabKey.watch(function tabKeyPress (event) {
            if ( isEnabled() && iface.isVisible() ) {
                var first = firstFocusable(iface.modalElem());
                var last = lastFocusable(iface.modalElem());

                var from = event.shiftKey ? first : last;
                if ( from === document.activeElement ) {
                    (event.shiftKey ? last : first).focus();
                    event.preventDefault();
                }
            }
        });
    }

    /** Manages setting the 'overflow: hidden' on the body tag */
    function manageBodyOverflow(iface, isEnabled) {
        var origOverflow;
        var body = new Elem(document.body);

        iface.beforeShow(function () {
            // Capture the current values so they can be restored
            origOverflow = body.elem.style.overflow;

            if (isEnabled()) {
                body.stylize({ overflow: "hidden" });
            }
        });

        iface.afterClose(function () {
            body.stylize({ overflow: origOverflow });
        });
    }

    /**
     * Displays a modal
     */
    return function picoModal(options) {

        if ( isString(options) || isNode(options) ) {
            options = { content: options };
        }

        var afterCreateEvent = observable();
        var beforeShowEvent = observable();
        var afterShowEvent = observable();
        var beforeCloseEvent = observable();
        var afterCloseEvent = observable();

        /**
         * Returns a named option if it has been explicitly defined. Otherwise,
         * it returns the given default value
         */
        function getOption ( opt, defaultValue ) {
            var value = options[opt];
            if ( typeof value === "function" ) {
                value = value( defaultValue );
            }
            return value === undefined ? defaultValue : value;
        }


        // The various DOM elements that constitute the modal
        var modalElem = build.bind(window, 'modal');
        var shadowElem = build.bind(window, 'overlay');
        var closeElem = build.bind(window, 'close');

        // This will eventually contain the modal API returned to the user
        var iface;


        /** Hides this modal */
        function forceClose () {
            shadowElem().hide();
            modalElem().hide();
            afterCloseEvent.trigger(iface);
        }

        /** Gracefully hides this modal */
        function close () {
            if ( beforeCloseEvent.trigger(iface) ) {
                forceClose();
            }
        }

        /** Wraps a method so it returns the modal interface */
        function returnIface ( callback ) {
            return function () {
                callback.apply(this, arguments);
                return iface;
            };
        }


        // The constructed dom nodes
        var built;

        /** Builds a method that calls a method and returns an element */
        function build ( name ) {
            if ( !built ) {
                var modal = buildModal(getOption, close);
                built = {
                    modal: modal,
                    overlay: buildOverlay(getOption, close),
                    close: buildClose(modal, getOption)
                };
                afterCreateEvent.trigger(iface);
            }
            return built[name];
        }

        iface = {

            /** Returns the wrapping modal element */
            modalElem: buildElemAccessor(modalElem),

            /** Returns the close button element */
            closeElem: buildElemAccessor(closeElem),

            /** Returns the overlay element */
            overlayElem: buildElemAccessor(shadowElem),

            /** Builds the dom without showing the modal */
            buildDom: returnIface(build),

            /** Returns whether this modal is currently being shown */
            isVisible: function () {
                return !!(built && modalElem && modalElem().isVisible());
            },

            /** Shows this modal */
            show: function () {
                if ( beforeShowEvent.trigger(iface) ) {
                    shadowElem().show();
                    closeElem();
                    modalElem().show();
                    afterShowEvent.trigger(iface);
                }
                return this;
            },

            /** Hides this modal */
            close: returnIface(close),

            /**
             * Force closes this modal. This will not call beforeClose
             * events and will just immediately hide the modal
             */
            forceClose: returnIface(forceClose),

            /** Destroys this modal */
            destroy: function () {
                modalElem().destroy();
                shadowElem().destroy();
                shadowElem = modalElem = closeElem = undefined;
            },

            /**
             * Updates the options for this modal. This will only let you
             * change options that are re-evaluted regularly, such as
             * `overlayClose`.
             */
            options: function ( opts ) {
                Object.keys(opts).map(function (key) {
                    options[key] = opts[key];
                });
            },

            /** Executes after the DOM nodes are created */
            afterCreate: returnIface(afterCreateEvent.watch),

            /** Executes a callback before this modal is closed */
            beforeShow: returnIface(beforeShowEvent.watch),

            /** Executes a callback after this modal is shown */
            afterShow: returnIface(afterShowEvent.watch),

            /** Executes a callback before this modal is closed */
            beforeClose: returnIface(beforeCloseEvent.watch),

            /** Executes a callback after this modal is closed */
            afterClose: returnIface(afterCloseEvent.watch)
        };

        manageFocus(iface, getOption.bind(null, "focus", true));

        manageBodyOverflow(iface, getOption.bind(null, "bodyOverflow", true));

        // If a user presses the 'escape' key, close the modal.
        escapeKey.watch(function escapeKeyPress () {
            if ( getOption("escCloses", true) && iface.isVisible() ) {
                iface.close();
            }
        });

        return iface;
    };

}));
